ptt.php

```diff
@@ -1,156 +1,179 @@
+		$post .= "Host: {$this->ip}\r\n";
+		$post .= "Authorization: Basic {$auth}\r\n";
+		$post .= "Connection: close\r\n";
+		$post .= "Content-Type: application/x-www-form-urlencoded\r\n";
+		$post .= "Content-Length: ".strlen($postData)."\r\n\r\n";
+		$post .= $postData;
+
+		$fp = @fsockopen ($this->ip, 80, $errno, $response, 10);
+		if($fp){
+			fputs($fp, $post);
+			flush();
+			while (!feof($fp)) {
+				$response .= fgets($fp, 1024);
+				flush();
+			}
+
+		}
+
+		return $response;
+	}
 }
-// Full-duplex, always on
-else {
-    echo "<Text>Press and release the [Unlock] soft key to return.</Text>";
-    echo "<SoftKeyItem>";
-    echo "<Name>Unlock</Name>";
-    echo "<URL>$URLBase/ptt.php?name=$MAC&amp;istalking=false</URL>";
-    echo "<Position>4</Position>";
-    echo "<URLDown>RTPTx:Stop</URLDown>";
-    echo "</SoftKeyItem>";
+
+class Push2Talk {
+	const XML_EXECUTE 		= '<CiscoIPPhoneExecute>%s</CiscoIPPhoneExecute>';
+	const XML_EXECUTE_ITEM 	= '<ExecuteItem Priority="%d" URL="%s"/>';
+	const URI_START			= 'RTPMRx:%s:%d';
+
+	public $multicastAddress= '225.3.15.13';
+	public $multicastPort 	= 20480;
+
+	var $authName 			= 'cisco';
+	var $authPassword 		= 'cisco';
+	var $deviceName 		= null;
+
+	var $devices 			= array();
+
+	function __construct(){ }
+
+	function setAuthentication($authName, $authPassword){
+		$this->authName 	= $authName;
+		$this->authPassword = $authPassword;
+	}
+
+	function addDevice($deviceIp){
+		$this->devices[] 	= new Device($deviceIp, $this->authName, $this->authPassword);
+	}
+
+	function addDevices(array $ips){
+		foreach($ips as $ip){
+			$this->addDevice($ip);
+		}
+	}
+
+	function start(){
+		$this->execute(array(
+			sprintf(Push2Talk::URI_START, $this->multicastAddress, $this->multicastPort)
+		));
+	}
+
+	function stop(){
+		$this->execute(array('RTPRx:Stop','RTPTx:Stop','Init:Services'));
+	}
+
+	function execute(array $uris, $priority = 0){
+		$xmlExecute = array_map(function ($value) {
+			return sprintf(Push2Talk::XML_EXECUTE_ITEM, $priority, $value);
+		}, $uris); 															// build ExecuteItem xml data-array
+		$xml = sprintf(Push2Talk::XML_EXECUTE, implode('',$xmlExecute));	// build xml data including ExecuteItem
+
+		foreach($this->devices as $device) {
+			$device->push($xml);
+		}
+	}
 }
 
-echo "</CiscoIPPhoneText>";
-?>
+
+$push2Talk = new Push2Talk;
+$push2Talk->setAuthentication('cisco', 'cisco');
+$push2Talk->addDevices(array('10.0.2.225', '10.0.2.227'));
+
+// optional settings
+//$push2Talk->multicastAddress = '225.3.15.13';
+//$push2Talk->multicastPort = 16384
+
+
+$response = '';
+do {  // while loop for error handling
+
+	$deviceName =  null;
+	$isTalking = false;
+	$action = isset($_GET['action']) ? $_GET['action'] : 'start';
+
+	if (!isset($_GET['name'])){
+		$response = '<CiscoIPPhoneText><Title>Error!</Title><Text>No MAC provided by phone!</Text></CiscoIPPhoneText>';
+		break;
+	}
+
+	$deviceName = $_GET['name'];
+
+	switch($action){
+		default:
```

Comment:
We should have an initial action which:
- Pushed the page to all the other participants
- After pushing the page, it can perform the push2talk->start() action (Should on start once, not every page refresh)